benchmark/utils/generator.hpp

```diff
@@ -0,0 +1,299 @@
+        gko::size_type stride)
+    {
+        auto res = Vec::create(exec, size, stride);
+        return res;
+    }
+
+    // creates a random multi_vector
+    static std::unique_ptr<Vec> create_multi_vector_random(
+        std::shared_ptr<const gko::Executor> exec, gko::dim<2> size)
+    {
+        auto res = Vec::create(exec);
+        res->read(gko::matrix_data<ValueType, itype>(
+            size,
+            std::uniform_real_distribution<gko::remove_complex<ValueType>>(-1.0,
+                                                                           1.0),
+            get_engine()));
+        return res;
+    }
+
+    static std::unique_ptr<Vec> initialize(
+        std::initializer_list<ValueType> il,
+        std::shared_ptr<const gko::Executor> exec)
+    {
+        return gko::initialize<Vec>(std::move(il), std::move(exec));
+    }
+};
+
+
+#if GINKGO_BUILD_MPI
+
+
+template <typename LocalGeneratorType>
+struct DistributedDefaultSystemGenerator {
+    using LocalGenerator = LocalGeneratorType;
+
+    using index_type = global_itype;
+    using local_index_type = typename LocalGenerator::index_type;
+    using value_type = typename LocalGenerator::value_type;
+
+    using Mtx = dist_mtx<value_type, local_index_type, index_type>;
+    using Vec = dist_vec<value_type>;
+
+    gko::matrix_data<value_type, index_type> generate_matrix_data(
+        rapidjson::Value& config) const
+    {
+        if (config.HasMember("filename")) {
+            std::ifstream in(config["filename"].GetString());
+            return gko::read_generic_raw<value_type, index_type>(in);
+        } else if (config.HasMember("stencil")) {
+            auto local_size = static_cast<global_itype>(
+                config["size"].GetInt64() / comm.size());
+            return generate_stencil<value_type, index_type>(
+                config["stencil"].GetString(), comm, local_size,
+                config["comm_pattern"].GetString() == std::string("optimal"));
+        } else {
+            throw std::runtime_error(
+                "No known way to generate matrix data found.");
+        }
+    }
+
+    std::shared_ptr<gko::LinOp> generate_matrix(
+        std::shared_ptr<gko::Executor> exec, rapidjson::Value& config) const
+    {
+        auto data = generate_matrix_data(config);
+        return generate_matrix_with_format(
+            std::move(exec), config["optimal"]["spmv"].GetString(), data);
+    }
+
+    std::shared_ptr<gko::LinOp> generate_matrix_with_format(
+        std::shared_ptr<gko::Executor> exec, const std::string& format_name,
+        const gko::matrix_data<value_type, index_type>& data,
+        rapidjson::Value* spmv_case = nullptr,
+        rapidjson::MemoryPoolAllocator<>* allocator = nullptr) const
+    {
+        auto part = gko::experimental::distributed::
+            Partition<itype, global_itype>::build_from_global_size_uniform(
+                exec, comm.size(), static_cast<global_itype>(data.size[0]));
```

Comment:
Maybe this Partition should be a parameter as well, allowing users to benchmark different kinds of partitions?

Reply:
I think instead of a function parameter, a better way to customize the partition would be to derive from the class, and override this function. I'm not sure how else the partition can be passed along.
So I will leave it in its current state.